File: UseCasesCore/Sources/UseCasesCore/Markers/RunAttestation.swift
import Crypto
import Foundation

/// Run attestation for the UNSIGNED verification-results ledger
/// (runAttestation.ts).
///
/// Every record `verify` writes carries an HMAC-SHA256 over its own canonical
/// JSON, keyed by 32 random bytes minted on this machine and kept outside the
/// repository, so a hand-typed line has no valid attestation. It is
/// tamper-EVIDENT, not tamper-proof: anyone who can read the key can forge one.
public enum RunAttestation {
  /// The record field carrying the attestation.
  public static let field = "run_attestation"

  private static let algorithm = "hmac-sha256"
  private static let keyByteCount = 32

  /// `hmac-sha256:<hex>` over the record's canonical JSON, the record's own
  /// attestation excluded. The key is read as node reads hex: leniently.
  public static func compute(
    record: JSONObject,
    key: String,
  ) throws(CodeUnitCanonicalJSONError) -> String {
    var payload = record
    payload[field] = nil
    let canonical = try CodeUnitCanonicalJSON.encode(.object(payload))
    let code = HMAC<SHA256>.authenticationCode(
      for: Data(canonical.utf8),
      using: SymmetricKey(data: NodeBuffer.hexDecode(key)),
    )
    return "\(algorithm):\(NodeBuffer.hexadecimal(code))"
  }

  /// True only for a record carrying the attestation this key produces. False,
  /// failing closed, when there is no well-formed key at all.
  public static func verify(
    record: JSONObject,
    key: String?,
  ) throws(CodeUnitCanonicalJSONError) -> Bool {
    guard let key, isRunKey(key),
          let claimed = record[field]?.stringValue, !claimed.isEmpty
    else {
      return false
    }
    let expected = try Array(compute(record: record, key: key).utf16)
    let claimedUnits = Array(claimed.utf16)
    guard claimedUnits.count == expected.count else {
      return false
    }
    var difference: UInt16 = 0
    for (left, right) in zip(claimedUnits, expected) {
      difference |= left ^ right
    }
    return difference == 0
  }

  /// The machine-local key, minted on first use. A missing, empty or malformed
  /// key file is REPLACED, never raised: a corrupt key degrades to "previous
  /// results read unattested", not to a crashed verify.
  public static func resolveLocalRunKey(
    keyPath: String,
    files: some TextFileStoring,
    mintKey: () -> String = mintRunKey,
  ) throws(FileAccessError) -> String {
    if let existing = try readLocalRunKey(keyPath: keyPath, files: files) {
      return existing
    }
    let minted = mintKey()
    try files.writeText("\(minted)\n", toPath: keyPath)
    return minted
  }

  /// The machine-local key, or nil when there is none. Never mints one: `scan`
  /// is strictly read-only.
  public static func readLocalRunKey(
    keyPath: String,
    files: some TextFileStoring,
  ) throws(FileAccessError) -> String? {
    guard let text = try files.readText(atPath: keyPath) else {
      return nil
    }
    let existing = JavaScriptString.trim(text)
    return isRunKey(existing) ? existing : nil
  }

  /// `UC_RUN_KEY_FILE` when it is set to something other than whitespace,
  /// trimmed; otherwise `<home>/.use-cases/run-key`, joined as `path.join`
  /// joins it.
  public static func defaultRunKeyPath(
    environment: [String: String],
    homeDirectory: String,
  ) -> String {
    if let override = environment["UC_RUN_KEY_FILE"] {
      let trimmed = JavaScriptString.trim(override)
      if !trimmed.isEmpty {
        return trimmed
      }
    }
    let joined = [homeDirectory, ".use-cases", "run-key"]
      .filter { !$0.isEmpty }
      .joined(separator: "/")
    return WorkspacePath.normalize(joined)
  }

  /// 32 random bytes, as lowercase hex.
  public static func mintRunKey() -> String {
    var generator = SystemRandomNumberGenerator()
    let bytes = (0 ..< keyByteCount).map { _ in
      UInt8.random(in: .min ... .max, using: &generator)
    }
    return NodeBuffer.hexadecimal(bytes)
  }

  /// `/^[0-9a-f]{64}$/`.
  static func isRunKey(_ key: String) -> Bool {
    let units = Array(key.utf16)
    return units.count == 64 && units.allSatisfy { unit in
      CodeUnits.isASCIIDigit(unit) || (0x61 ... 0x66).contains(unit)
    }
  }
}

File: packages/core/src/markers/runAttestation.ts
// Run attestation for the UNSIGNED verification-results ledger.
//
// THE PROBLEM. `.use-cases/verification-results.jsonl` is a plain text file that
// `verify --out` writes and `scan` reads. `scan` used to trust any line in it
// whose hashes matched the current code, which meant the keyless acceptance
// claim rewarded typing over running: appending one hand-written line moved the
// number, while genuinely driving the product moved nothing.
//
// THE FIX. Every record `verify` writes carries an HMAC over its own content,
// keyed by a secret that lives on the machine and NOT in the repo. Only the tool
// that ran the verifier holds it, so a typed line has no valid attestation and
// `scan` demotes it to UNATTESTED_LOCAL.
//
// WHAT THIS IS NOT. It is not the signing tier. There is no CI, no keyring, no
// key rotation and nothing for the user to do: the key is 32 random bytes minted
// on first use. It is tamper-EVIDENT, not tamper-PROOF — anyone who can read the
// key file can forge a record. That is the honest bound of a keyless local tier,
// and it is a large step up from "any line in a tracked file is proof".
//
// WHY THE KEY LIVES OUTSIDE THE REPO. The results ledger is transient, per-
// machine output — `use-cases init` gitignores it for exactly that reason. A key stored
// beside it would either be committed (and so forgeable by anyone with the repo)
// or need its own ignore entry in every adopting project. Keeping it under the
// user's home makes "this machine ran it" the literal meaning of the
// attestation: a results ledger someone else committed reads unattested here,
// which is correct — their run is not your evidence.
import { createHmac, randomBytes } from "node:crypto";
import { homedir } from "node:os";
import { join } from "node:path";
import { canonicalJson } from "./canonicalJson.js";
import type { MarkerFs } from "./cli/io.js";

// The field carrying the attestation on a verification-result record.
export const RUN_ATTESTATION_FIELD = "run_attestation";

// Named in the value so a reader of the ledger can tell what the string is.
const RUN_ATTESTATION_ALG = "hmac-sha256";

// 32 bytes, hex. Long enough that guessing is not a strategy.
const RUN_KEY_BYTES = 32;
const RUN_KEY_PATTERN = /^[0-9a-f]{64}$/;

// Compute the attestation for one verification-result record. The record's own
// `run_attestation` field is excluded (it cannot cover itself), and the input is
// canonical JSON, so key ORDER never changes the value — the ledger is re-sorted
// and rewritten on every merge, and an honest record must survive that.
export function computeRunAttestation(record: Record<string, unknown>, key: string): string {
  const payload = { ...record };
  delete payload[RUN_ATTESTATION_FIELD];
  const mac = createHmac("sha256", Buffer.from(key, "hex"))
    .update(canonicalJson(payload))
    .digest("hex");
  return `${RUN_ATTESTATION_ALG}:${mac}`;
}

// Does this record carry an attestation this machine could have written? False
// for a record with no attestation, a tampered one, one minted elsewhere, and —
// fail closed — for every record when no key could be resolved at all.
export function verifyRunAttestation(
  record: Record<string, unknown>,
  key: string | null
): boolean {
  if (key === null || !RUN_KEY_PATTERN.test(key)) {
    return false;
  }
  const claimed = record[RUN_ATTESTATION_FIELD];
  if (typeof claimed !== "string" || claimed === "") {
    return false;
  }
  const expected = computeRunAttestation(record, key);
  // Both sides are fixed-length hex we produced, so a length mismatch is a
  // malformed claim rather than a timing signal worth protecting.
  if (claimed.length !== expected.length) {
    return false;
  }
  let diff = 0;
  for (let index = 0; index < expected.length; index += 1) {
    diff |= claimed.charCodeAt(index) ^ expected.charCodeAt(index);
  }
  return diff === 0;
}

// Read the machine-local run key, minting one on first use. A missing OR
// unreadable OR malformed file is replaced rather than raised: a corrupt key
// must degrade to "previous results read unattested, re-run verify", never to a
// crashed verify.
export function resolveLocalRunKey(
  keyPath: string,
  fs: MarkerFs,
  mintKey: () => string = () => randomBytes(RUN_KEY_BYTES).toString("hex")
): string {
  const existing = fs.readText(keyPath)?.trim();
  if (existing !== undefined && RUN_KEY_PATTERN.test(existing)) {
    return existing;
  }
  const minted = mintKey();
  fs.writeText(keyPath, `${minted}\n`);
  return minted;
}

// The machine-local run key, or null when no key exists yet and none may be
// minted. `scan` is strictly read-only (its whole contract), so it must never
// create a key: on a machine that has never run `verify`, every result correctly
// reads unattested.
export function readLocalRunKey(keyPath: string, fs: MarkerFs): string | null {
  const existing = fs.readText(keyPath)?.trim();
  return existing !== undefined && RUN_KEY_PATTERN.test(existing) ? existing : null;
}

// Where the machine-local run key lives by default: under the USER'S HOME, not
// the repo. `UC_RUN_KEY_FILE` overrides it — a container that wants the key on a
// mounted volume, or a test that must not touch the real home, sets that.
export function defaultRunKeyPath(env: NodeJS.ProcessEnv = process.env): string {
  const override = env.UC_RUN_KEY_FILE;
  if (typeof override === "string" && override.trim() !== "") {
    return override.trim();
  }
  return join(homedir(), ".use-cases", "run-key");
}
